add State to handler

chriswoodrich opened the pull request.

chriswoodrich commented on lib/lita/handler/state.rb:
```diff
@@ -0,0 +1,49 @@
+      # @since 5.0.0 ?
+      attr_reader :store, :lock
+
+      # Instatiates a State object and it's internal dependencies.
```
ಠ__ಠ  Aren't you married to an English teacher?

jimmycuadra replied:
Owned.

jimmycuadra commented:
To do:

* The state class itself should just be `Lita::State`.
* `Lita::Robot` should have an instance of the state class upon initialization.
* `Lita::Handler::Common` should include a convenience method for accessing the the current robot's state object with the handler's namespace as the key.
* Remove the `state` instance method from the state class. The data structure itself shouldn't be public.
* The `synchronize` method should be private.
* Remove the question mark from the "since" doc tag. This will definitely be 5.0.0.
* Don't use any sort of eval in tests. Stick to the public interface.

chriswoodrich commented:
@jimmycuadra addressed your comments except for the 3rd one...Do you mean every handler should have it's own ```State```?  I don't get your comment "...with the handler's namespace as the key."

jimmycuadra commented on lib/lita/handler/common.rb:
```diff
@@ -72,10 +72,15 @@ def log_error(robot, error)
 
+      # The current unpersisted and shared {State} instance.
+      # @return [State]
+      attr_reader :state
```
Don't store or expose the state object directly. Initialize the value for the handler with `robot.state.set(self.class.namespace, {})` and add an instance method called `state` that does `robot.state.get(self.class.namespace)`.

jimmycuadra commented on lib/lita/robot.rb:
```diff
@@ -59,11 +65,12 @@ class Robot
     def initialize(registry = Lita)
       @registry = registry
-      @name = config.robot.name
+      @name         = config.robot.name
```
Don't change the alignment of these assignments, plehz.

chriswoodrich replied:
:(

jimmycuadra commented on lib/lita/state.rb:
```diff
@@ -0,0 +1,43 @@
+
+module Lita
+  class State
+    # A handler class that gives in-memory, non-persited, object storage
```
It's not a handler class. (Also missing a period.)

jimmycuadra commented on lib/lita/state.rb:
```diff
@@ -0,0 +1,43 @@
+  class State
+    # A handler class that gives in-memory, non-persited, object storage
+    # @since 5.0.0
+    attr_reader :store, :lock
```
Either move these readers below the `private` keyword or just use instance variables. This stuff shouldn't be exposed publicly.

jimmycuadra commented on spec/lita/handler/state_spec.rb:
```diff
@@ -0,0 +1,42 @@
+require "spec_helper"
+
+describe Lita::State do
+  describe "#initialize" do
```
You can remove this example—it doesn't provide any value.

jimmycuadra commented on spec/lita/handler/state_spec.rb:
```diff
@@ -0,0 +1,42 @@
+    end
+  end
+
+  describe "#synchronize" do
```
Don't test private methods.

jimmycuadra commented on spec/lita/handler/state_spec.rb:
```diff
@@ -0,0 +1,42 @@
+  before :each do
+    dat = data
+    storage.set(:expensive_key, dat)
+  end
```
All this setup is not necessary—let's just have one example for `#get` and one example for `#set`. (Remove the let blocks and the before block.)

jimmycuadra commented:
Getting pretty close! Thanks for your work on this. :}

jimmycuadra commented on lib/lita/state.rb:
```diff
@@ -0,0 +1,43 @@
+    # @since 5.0.0
+    attr_reader :store, :lock
+
+    # Instantiate a State object and it's internal dependencies.
```
Grammar!

jimmycuadra commented on lib/lita/state.rb:
```diff
@@ -0,0 +1,43 @@
+
+    private
+
+    # Gives the handler threadsafe
```
Full YARD isn't necessary for private methods. Just a single sentence description of its purpose is fine.

jimmycuadra commented on lib/lita/state.rb:
```diff
@@ -0,0 +1,43 @@
+module Lita
+  class State
+    # A handler class that gives in-memory, non-persited, object storage
+    # @since 5.0.0
```
You can move both the class description and since tag above the class declaration and then it will apply to the whole class, so you can remove the since tags from the individual methods.

jimmycuadra commented:
Replaced by #178.
